Modules/_interpretersmodule.c

```diff
@@ -0,0 +1,1892 @@
+    char *msg;
+};
+
+static struct _shared_exception *
+_get_shared_exception(void)
+{
+    struct _shared_exception *err = PyMem_NEW(struct _shared_exception, 1);
+    if (err == NULL)
+        return NULL;
+    PyObject *exc;
+    PyObject *value;
+    PyObject *tb;
+    PyErr_Fetch(&exc, &value, &tb);
+    PyObject *msg;
+    if (value == NULL)
+        msg = PyUnicode_FromFormat("%S", exc);
+    else
+        msg = PyUnicode_FromFormat("%S: %S", exc, value);
+    if (msg == NULL) {
+        err->msg = "unable to format exception";
+        return NULL;
+    }
+    err->msg = (char *)PyUnicode_AsUTF8(msg);
+    if (err->msg == NULL) {
+        err->msg = "unable to encode exception";
+    }
+    return err;
+}
+
+static PyObject * RunFailedError;
+
+static int
+interp_exceptions_init(PyObject *ns)
+{
+    // XXX Move the exceptions into per-module memory?
+
+    // An uncaught exception came out of interp_run_string().
+    RunFailedError = PyErr_NewException("_interpreters.RunFailedError",
+                                        PyExc_RuntimeError, NULL);
+    if (RunFailedError == NULL)
+        return -1;
+    PyDict_SetItemString(ns, "RunFailedError", RunFailedError);
+
+    return 0;
+}
+
+static void
+_apply_shared_exception(struct _shared_exception *exc)
+{
+    PyErr_SetString(RunFailedError, exc->msg);
+}
+
+/* channel-specific code */
+
+static PyObject *ChannelError;
+static PyObject *ChannelNotFoundError;
+static PyObject *ChannelClosedError;
+static PyObject *ChannelEmptyError;
+
+static int
+channel_exceptions_init(PyObject *ns)
+{
+    // XXX Move the exceptions into per-module memory?
+
+    // A channel-related operation failed.
+    ChannelError = PyErr_NewException("_interpreters.ChannelError",
+                                      PyExc_RuntimeError, NULL);
+    if (ChannelError == NULL)
+        return -1;
+    PyDict_SetItemString(ns, "ChannelError", ChannelError);
+
+    // An operation tried to use a channel that doesn't exist.
+    ChannelNotFoundError = PyErr_NewException("_interpreters.ChannelNotFoundError",
+                                              ChannelError, NULL);
+    if (ChannelNotFoundError == NULL)
+        return -1;
+    PyDict_SetItemString(ns, "ChannelNotFoundError", ChannelNotFoundError);
+
+    // An operation tried to use a closed channel.
+    ChannelClosedError = PyErr_NewException("_interpreters.ChannelClosedError",
+                                            ChannelError, NULL);
+    if (ChannelClosedError == NULL)
+        return -1;
+    PyDict_SetItemString(ns, "ChannelClosedError", ChannelClosedError);
+
+    // An operation tried to pop from an empty channel.
+    ChannelEmptyError = PyErr_NewException("_interpreters.ChannelEmptyError",
+                                           ChannelError, NULL);
+    if (ChannelEmptyError == NULL)
+        return -1;
+    PyDict_SetItemString(ns, "ChannelEmptyError", ChannelEmptyError);
+
+    return 0;
+}
+
+struct _channelend;
```

Comment:
Maybe typedef it so that it's easier to use?

Reply:
fixed